src/main/java/io/airlift/units/Count.java

```diff
@@ -0,0 +1,205 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.units;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static io.airlift.units.Preconditions.checkArgument;
+import static java.lang.Long.parseLong;
+import static java.lang.Math.floor;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class Count
+        implements Comparable<Count>
+{
+    private static final Pattern PATTERN = Pattern.compile("^\\s*(\\d+)\\s*([a-zA-Z]?)\\s*$");
+
+    // We iterate over the MAGNITUDES constant in convertToMostSuccinctRounded()
+    // instead of Magnitude.values() as the latter results in non-trivial amount of memory
+    // allocation when that method is called in a tight loop. The reason is that the values()
+    // call allocates a new array at each call.
+    private static final Magnitude[] MAGNITUDES = Magnitude.values();
+
+    /**
+     * @return count with no bigger value than 1000 in succinct magnitude, fractional part is rounded
+     */
+    public static Count succinctRounded(long count)
+    {
+        return succinctRounded(count, Magnitude.SINGLE);
+    }
+
+    /**
+     * @return count with no bigger value than 1000 in succinct magnitude, fractional part is rounded
+     */
+    public static Count succinctRounded(long count, Magnitude magnitude)
+    {
+        return new Count(count, magnitude).convertToMostSuccinctRounded();
+    }
+
+    private final long value;
+    private final Magnitude magnitude;
+
+    public Count(long count, Magnitude magnitude)
+    {
+        checkArgument(!Double.isInfinite(count), "count is infinite");
+        checkArgument(!Double.isNaN(count), "count is not a number");
+        checkArgument(count >= 0, "count is negative");
+        requireNonNull(magnitude, "magnitude is null");
+
+        this.value = count;
+        this.magnitude = magnitude;
+    }
+
+    public long getValue()
+    {
+        return value;
+    }
+
+    public Magnitude getMagnitude()
+    {
+        return magnitude;
+    }
+
+    public long getValue(Magnitude magnitude)
+    {
+        if (value == 0L) {
+            return 0L;
+        }
+
+        long scale = this.magnitude.getFactor() / magnitude.getFactor();
+        if (scale == 0) {
+            throw new IllegalArgumentException(format("Unable to return value %s in magnitude %s due precision loss", this, magnitude));
+        }
+        return value * scale;
+    }
+
+    public Count convertTo(Magnitude magnitude)
+    {
+        requireNonNull(magnitude, "magnitude is null");
```

Comment:
Move this check to `getValue` (or remove it since the caller will get an NPE anyway)